docs/src/filters/writing_custom_filters.md

```diff
@@ -38,109 +38,91 @@ These components come together to form the [filter chain].
-       fn name(&self) -> &'static str {
-           // We provide the name of filter that we defined earlier.
-           NAME
-       }
-       fn create_filter(&self, _: CreateFilterArgs) -> Result<Box<dyn Filter>, Error> {
-           Ok(Box::new(Greet))
-       }
-   }
-   ```
-
-1. **Start the proxy**
-
-   We can run the proxy in the exact manner as the default Quilkin binary using the [run][runner::run] function, passing in our custom [FilterFactory].
-   Lets add a main function that does that. Quilkin relies on the [Tokio] async runtime so we need to import that crate and wrap our main function with it.
-
-   Add Tokio as a dependency in `Cargo.toml`.
-   ```toml
-   [dependencies]
-   quilkin = "0.1.0-dev"
-   tokio = { version = "1", features = ["full"]}
-   ```
-
-   Add a main function that starts the proxy.
-   ```rust, no_run
-   // src/main.rs
-   # use quilkin::filters::{CreateFilterArgs, Filter, Error, FilterFactory};
-   # struct GreetFilterFactory;
-   # impl FilterFactory for GreetFilterFactory {
-   #     fn name(&self) -> &'static str {
-   #         "greet.v1"
-   #     }
-   #     fn create_filter(&self, _: CreateFilterArgs) -> Result<Box<dyn Filter>, Error> {
-   #         unimplemented!()
-   #     }
-   # }
-   use quilkin::{filters::DynFilterFactory, run};
-
-   #[tokio::main]
-   async fn main() {
-       run(vec![Box::new(GreetFilterFactory) as DynFilterFactory]).await.unwrap();
-   }
-   ```
+#### 1. Import the Quilkin crate
+
+```bash
+# Start with a new crate
+cargo new --bin quilkin-filter-example
+```
+Add Quilkin as a dependency in `Cargo.toml`.
+```toml
+[dependencies]
+quilkin = "0.2.0"
+```
+
+#### 2. Implement the filter traits
+
+It's not terribly important what the filter in this example does so let's write a `Greet` filter that appends `Hello` to every packet in one direction and `Goodbye` to packets in the opposite direction.
+
+We start with the [Filter] implementation
+
+```rust,no_run,noplayground
+struct Greet;
+
+impl Filter for Greet {
+    fn read(&self, mut ctx: ReadContext) -> Option<ReadResponse> {
+        ctx.contents.splice(0..0, String::from("Hello ").into_bytes());
+        Some(ctx.into())
+    }
+    fn write(&self, mut ctx: WriteContext) -> Option<WriteResponse> {
+        ctx.contents.splice(0..0, String::from("Goodbye ").into_bytes());
+        Some(ctx.into())
+    }
+}
+```
+
+Next, we implement a [FilterFactory] for it and give it a name:
+
+```rust,no_run,noplayground
+pub const NAME: &str = "greet.v1";
+
+pub fn factory() -> DynFilterFactory {
+    Box::from(GreetFilterFactory)
+}
+
+struct GreetFilterFactory;
+impl FilterFactory for GreetFilterFactory {
+    fn name(&self) -> &'static str {
+        NAME
+    }
+    fn create_filter(&self, args: CreateFilterArgs) -> Result<Box<dyn Filter>, Error> {
+        Ok(Box::new(Greet))
+    }
+}
+```
+
+> As a convention we implement a `pub factory() -> DynFilterFactory` function, to 
+> make the next integration step easier.
```

Comment:
I don't think this is correct? the `factory()` function or constructor wouldn't be visible at all from user code and what's passed into quilkin is the `DynFilterFactory` so it doesn't matter how the object is generated?
```suggestion
```

Reply:
Just so I'm clear, are you saying the description isn't correct? 

If I look at all our Filter implementations, they all have a `pub factory() -> DynFilterFactory` - so I thought it would be good to call out that we conventionally do that, and why.

You are 100% correct that the end use could create a constructor any way, but I thought it would be good to show our usual convention. Maybe the description should be:

> As a convention we generally implement a `pub factory() -> DynFilterFactory` function as a constructor, but a Factory 
> can be constructed as you see fit.

To explain that while you may see this common pattern, it's more by convention, rather than technical limitation. WDYT?

Reply:
Ah, the incorrect part was mostly the 'to make the next integration step easier' bit since the function itself didn't matter.

I think in this case the fact that the quilkin codebase used a function called factory vs anything else doesn't matter at all so it can add confusion to call it out in the first place (as a user I'd find myself reading into what that means and quickly realise it doesn't matter which is wasted time), the user would declare and call the function within their own crate so the function can be called anything or they could construct the factory by some other means (I don't see it as a convention here since both crates don't need to worry about it in the first place).

Reply:
That's fair. Sounds good, I'll remove the comment 👍🏻 

Worst case, if people start asking, we can add it back in 😄